### DISCUSSION TITLE
Basic Integration Questions

### DISCUSSION THREAD

**Discussion by von-Chaps** [ekutner/home-connect-hass#75] [Community Support]
Here is a screenshot of my current layout of my oven entities. I've hidden some of the ones I don't use like the child lock, etc, but these are the ones I'm interested in plus some of the connectivity state ones which I've been observing whilst I try and understand what's going on. I have set readable names for the entities, but I think it's still clear which is which.

This screenshot shows the oven after it's been running a program for about 10 mins. I set the program in the HC app not in here.

![Screenshot_20220627-173811_Vivaldi Browser](https://user-images.githubusercontent.com/53400820/176040861-a53230ab-143b-4098-941b-d489c61551e8.jpg)

I have a whole bunch of questions that I have not figured out. If anyone has any insight on any of them, I'd be interested to hear.

1. Why does "Selected Program" show as "Unknown" when the program is "Hot Air" in the input box?
2. Why is "Heat Up Progress" 0% when the temperature shows this is not the case?
3. The duration is correct at 1500 secs, but why is the "Elapsed Program Time" zero, when it should be around 10 mins?
4. Why is the "Program Progress" 0% - it should be around 20%
5. Why does "Remaining Program Time" say "In 20 mins"?

I tried changing the duration and setpoint temp with the program running, but neither control results in any change on the oven.

I should say that all this stuff shows correctly in the HC app itself. It's almost as if the relevant info is not being sent to the integration since it didn't start the program. Could that be? I used the HC app to start this program because I can't get the setpoint or duration to change from within HA, but that's a different question.

**Comment by ekutner** [comment]
> 1. Why does "Selected Program" show as "Unknown" when the program is "Hot Air" in the input box?

The program is active so it is displayed in the "Active Program" sensor. This is reflecting how HC reports the state of the appliance. At this time there is no selected program, only an active program.

For all the other questions the answer is the same. The integration doesn't do any polling, it relies on events received from HC to update its state. If values are not updated than that is, with 99% certainty, because HC didn't send an event that updates them. Getting to 100% requires going through the logs but as noted, it is almost certainly the case.

The HC mobile app is using a different API that the public one used by the integration. In fact, it can even communicate directly with the appliances over your local home network, without even going out to the cloud, so it just works very differently. A first debugging step is to turn off the wifi on your phone so the app is forced to communicate through the cloud service, in my experience that at least removes some of the differences in behavior, although not all.

**Comment by von-Chaps** [comment]
As a general point, neither my HA instance/app nor the HC app get to access my oven directly on the LAN. That's because the oven is on a different VLAN from my control points. So, everything goes through the HC servers. I am sure they do still use different APIs since they wouldn't want people producing "better" apps than they do. Rate limiting comes to mind. I'll bet the HC app is not rate limited!

On point 1 about the "Selected Program", I struggled to think of a case where this would be different from "Active Program" and the only thing I could think of was the case of a delayed start. However, this is not the case. A program becomes active as soon as I set it up even if it has delayed start. As such, I can not find a case where "Selected Program" is anything other than "Unavailable". As such, I'm going to hide that entity as it's not of any use to me. Of course, that may be just my use cases and my oven.

As for the other stuff, I understand what you're saying about what comes back from HC. I'll need to look into that in more detail as, on first glance, different programs result in different data being returned which seems remarkably inconsistent for a published API. I'll have more on this later, but I will use your debug mode to observe what comes back from the API calls before bothering you with it here.

**Comment by ekutner** [comment]
In appliances with start/stop buttons the selected and active programs are two separate things. For example on a washing machine you first select the program and maybe set some program options like spin speed and temperature, at that point there is a selected program but no active program. When you push the start button, the selected program becomes active.

**Comment by von-Chaps** [comment]
Yea, that's a totally cromulent answer and is what I was thinking. Except that, of course, the integration _does_ offer start and stop options and, indeed, as far as I can see that is the _only_ way to start and stop programs on my oven using the integration. Thus, I thought "selected" would reflect the program I had chosen in the drop down selector, but before I had pressed "start" at which point it would move to "Active Program". This is not the case of course. I do understand that the start and stop in the integration are not really what you are referring to, but it does drive the user's thinking in that direction.

In this particular case, it does seem that "Selected Program" is not applicable to my device. And this leads to the real issue: _which of the entities surfaced by the integration really apply to the appliance in question_? Another example might be trying to operate a coffee machine and being presented with a "Duration" option - which I assume is meaningless. Surely it creates confusion and generates questions (like all mine) when people are presented with options and settings that don't apply. It is also difficult to know if an entity genuinely doesn't apply, represents a feature the user hasn't discovered or even is a bug in the integration. Another example is that I can't get "duration" and "setpoint temperature" to do anything at all to my oven. At this point, I don't know if they should work or not, but I'll bring this up as a separate issue once I've looked into it more.

Now, I understand how you've gone about the implementation and think it's correct. However, generally a decent API would allow one to query device capabilities. It seems HC doesn't provide this feature and just throws everything at you regardless of whether it applies to the appliance(s) in question. That has to create difficulties.

I am starting to wonder if there shouldn't be GUI components that pick from the set of available entities rather like the climate one. There would be one for oven, another for dishwasher, etc and they would only present applicable entities and in a way that reflects the device to some extent. Easy to say I guess - just throwing it out there.

Just thinking aloud, there is a theoretical difference between an entity being "Unavailable" in HA and in HC. In HA, it means it could not be populated for some reason (e.g. offline). In HC it means "not applicable to this device". This distinction is further blurred because of the lazy loading you have had to do to try and deal with the ridiculous rate limits.

I'd be interested in hearing what you think if you have the inclination, but I don't feel I'm actually asking you a specific question here so no worries if you are fed up with my wordiness!

**Comment by ekutner** [comment]
It is really very simple. The integration only creates entities which are applicable to the appliance, if you have a Duration entity then that information is provided by the API for your appliance. The selected program is a part of the API and it being null *IS* a relevant piece of information, feel free to ignore it if it's not relevant to you.
The start button is there because the API always needs an explicit start action and therefor the integration always works the same way. First you select a program in the select a program then start it. For programs that can't be selected the integration stores the selected program and then applies it when you click the start button.

**Comment by von-Chaps** [comment]
OK, well in that case I really am confused :)

Up thread, you said that "Selected Program" was for devices with start/stop buttons where you could pre-select a program, set some options and then start it. My device does not have start/stop buttons nor does it have any ability to pre-select a program. I can't see under what circumstances "Selected Program" can be anything other than null.

Either;
1. My understanding of how to operate the device is incomplete (always possible)
2. This item is being returned by the API even though it does not apply to my device
3. Something is broken

All of my above comments really boil down to my being able to tell which of these 3 is the case for any entities I don't seem to be able to use.

Another example: I've been unable to use "startinrelative" to delay the start time. I've dug into it a bit and it seems to be mostly a case of "3. Something is broken", but it could easily turn out to be 1 or 2.

**Comment by ekutner** [comment]
It's a combination of 1 & 2.
The sensor reports the response of the .../selected program endpoint, which for you appliance is null. The operation model is as I described it.

**Comment by von-Chaps** [comment]
OK. Many thanks for all your patient support and explanations. I think the best thing I can do is go off and take a good look at the API to try and get my head around why some things work and some don't.

Thanks for the great work and sorry to have been a bit of a pain!

**Comment by esvane** [comment]
First of all, let me tell you that I love your work. The integration works very well. and I am already using it in a lot of automations. 

The only thing is, i'm not able to fix something in the `bsh.common.option.duration` on my oven. It expects a value with a minimum of `1`. The way I normally use my oven is without setting that timer. But when I dont set a duration, it only uses the last given value.  It seems it cannot be set to `0` or be disabled. Starting the oven manually, I can skip this option. Is it possible to start the oven without setting a value with a minimun of `1`?

**Comment by ekutner** [comment]
@esvane Sorry, but I didn't understand the problem. Please open an issue about it, with more details explanation of the problem and logs.
